File: sample_apps_reinvent2021/python/utils/TimestreamDependencyHelper.py
import boto3
import json


class TimestreamDependencyHelper:

    def __init__(self, region):
        self.region = region
        self.sns_client = boto3.client('sns', region_name=region)
        self.iam_client = boto3.client('iam', region_name=region)
        self.sqs_client = boto3.client('sqs', region_name=region)
        self.account_id = boto3.client("sts").get_caller_identity()["Account"]
        self.s3_client = boto3.client('s3', region_name=region)
        self.s3_resource = boto3.resource('s3')

    def create_sns(self, topic_name):
        print("\nCreating sns topic")
        try:
            response = self.sns_client.create_topic(Name=topic_name)
            print("Sns topic created successfully, topic arn:", response['TopicArn'])
            return response['TopicArn']
        except Exception as err:
            print("Failed creating sns: ", err)
            raise err

    def delete_sns(self, topic_arn):
        print("Deleting sns topic")
        try:
            self.sns_client.delete_topic(TopicArn=topic_arn)
            print("Sns topic deleted successfully")
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed deleting sns: ", err)

    def create_sqs_queue(self, queue_name):
        print("\nCreating sqs queue")
        try:
            response = self.sqs_client.create_queue(QueueName=queue_name)
            print("Sqs queue created successfully, queue url:", response['QueueUrl'])
            return response['QueueUrl']
        except Exception as err:
            print("Failed creating sqs: ", err)
            raise err

    def get_queue_arn(self, queue_url):
        try:
            response = self.sqs_client.get_queue_attributes(QueueUrl=queue_url,
                                                            AttributeNames=['QueueArn'])
            return response['Attributes']['QueueArn']
        except Exception as err:
            print("Failed getting sqs queue arn: ", err)
            raise err

    def set_sqs_access_policy(self, queue_url, queue_arn, topic_arn):
        access_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Sid": "topic-subscription-" + topic_arn,
                    "Effect": "Allow",
                    "Principal": {
                        "AWS": "*"
                    },
                    "Action": "sqs:SendMessage",
                    "Resource": queue_arn,
                    "Condition": {
                        "ArnLike": {
                            "aws:SourceArn": topic_arn
                        }
                    }
                }
            ]
        }
        access_policy = json.dumps(access_policy)
        try:
            self.sqs_client.set_queue_attributes(QueueUrl=queue_url, Attributes={'Policy': access_policy})
        except Exception as err:
            print("Failed setting sqs queue policy: ", err)
            raise err

    def subscribe_to_sns_topic(self, topic_arn, queue_arn):
        print("Subscribing queue to sns topic")
        try:
            response = self.sns_client.subscribe(TopicArn=topic_arn,
                                                 Protocol='sqs',
                                                 Endpoint=queue_arn,
                                                 ReturnSubscriptionArn=True)
            print("Successfully subscribed to sns topic, subscription arn is :", response['SubscriptionArn'])
            return response['SubscriptionArn']
        except Exception as err:
            print("Failed subscribing to sns: ", err)
            raise err

    def unsubcribe_from_sns(self, subscription_arn):
        try:
            self.sns_client.unsubscribe(SubscriptionArn=subscription_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed unsubscribing from sns: ", err)

    def receive_message(self, queue_url):
        try:
            response = self.sqs_client.receive_message(QueueUrl=queue_url,
                                                       WaitTimeSeconds=20)
            if 'Messages' in response and response['Messages']:
                response = response['Messages'][0].get('Body', '{}')
                response = json.loads(response)
                return response
            return None
        except Exception as err:
            print("Error while receiving message: ", err)
            raise err

    def delete_sqs_queue(self, queue_url):
        print("\nDeleting sqs queue")
        try:
            self.sqs_client.delete_queue(QueueUrl=queue_url)
            print("Successfully deleted queue")
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed deleting queue: ", err)

    def create_role(self, role_name, stage, region):
        print("\nCreating IAM role for scheduled query")
        assumed_role_policy_document = {
            'Version': '2012-10-17',
            'Statement': [
                {
                    'Effect': 'Allow',
                    'Principal': {
                        "Service": self.get_service_name(stage, region)
                    },
                    'Action': 'sts:AssumeRole'
                }
            ]
        }
        assumed_role_policy_document = json.dumps(assumed_role_policy_document)
        try:
            response = self.iam_client.create_role(Path='/',
                                                   RoleName=role_name,
                                                   AssumeRolePolicyDocument=assumed_role_policy_document
                                                   )
            print("Successfully created IAM role for accessing scheduled query: ", response['Role']['Arn'])
            return response['Role']['Arn']
        except self.iam_client.exceptions.EntityAlreadyExistsException:
            print("Role already exists")
            return f"arn:aws:iam::{self.account_id}:role/{role_name}"
        except Exception as err:
            print("Failed creating role: ", err)
            raise err

    def delete_role(self, role_arn):
        print("Deleting role: ", role_arn)
        try:
            self.iam_client.delete_role(RoleName=role_arn)
            print("Successfully deleted role: ", role_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed deleting role: ", err)

    def create_policy(self, policy_name):
        print("Creating policy for Scheduled Query access")
        policy_document = {
            'Version': '2012-10-17',
            'Statement': [
                {
                    'Action': [
                        'kms:Decrypt',
                        'sns:Publish',
                        'timestream:describeEndpoints',
                        'timestream:Select',
                        'timestream:SelectValues',
                        'timestream:WriteRecords',
                        's3:GetObject',
                        's3:List*',
                        's3:Put*',
                    ],
                    'Resource': '*',
                    'Effect': 'Allow'
                }
            ]
        }
        policy_document = json.dumps(policy_document)
        try:
            response = self.iam_client.create_policy(PolicyName=policy_name,
                                                     PolicyDocument=policy_document)
            print("Successfully created policy: ", response['Policy']['Arn'])
            return response['Policy']['Arn']
        except self.iam_client.exceptions.EntityAlreadyExistsException as e:
            print("Policy already exists")
            return f"arn:aws:iam::{self.account_id}:policy/{policy_name}"
        except Exception as err:
            print("Failed creating policy: ", err)
            raise err

    def delete_policy(self, policy_arn):
        print("Deleting policy")
        try:
            self.iam_client.delete_policy(PolicyArn=policy_arn)
            print("Successfully deleted policy: ", policy_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed deleting policy: ", err)

    def attach_policy_to_role(self, role_name, policy_arn):
        print("Attaching policy to role: ", role_name)
        try:
            self.iam_client.attach_role_policy(RoleName=role_name,
                                               PolicyArn=policy_arn)
            print("Successfully attached policy to role")
        except Exception as err:
            print("Failed attaching policy to role: ", err)
            raise err

    def detach_policy_from_role(self, role_name, policy_arn):
        print("\nDetaching policy from role: ", role_name)
        try:
            self.iam_client.detach_role_policy(RoleName=role_name,
                                               PolicyArn=policy_arn)
            print("Successfully detached policy from role")
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed detaching policy from role: ", err)

    def get_service_name(self, stage, region):
        if stage == "prod":
            return "timestream.{}.amazonaws.com".format(region)
        else:
            return "{}.{}.timestream.aws.internal".format(region, stage)

    def create_s3_bucket(self, bucket_name):
        print("\nCreating S3 bucket")
        try:
            if self.region == 'us-east-1':
                self.s3_client.create_bucket(Bucket=bucket_name)
            else:
                self.s3_client.create_bucket(Bucket=bucket_name,
                                             CreateBucketConfiguration={'LocationConstraint': self.region})
            print("Successfully created S3 bucket: ", bucket_name)
            return bucket_name
        except Exception as err:
            print("Failed creating S3 bucket: ", err)
            raise err

    def delete_s3_bucket(self, bucket_arn):
        print("Deleting S3 bucket")
        try:
            bucket = self.s3_resource.Bucket(bucket_arn)
            bucket.objects.all().delete()
            bucket.delete()
            print("Successfully deleted S3 bucket: ", bucket_arn)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed to delete S3 bucket: ", err)

    def list_s3_objects(self, bucket_name, prefix):
        print("\nListing object in S3 bucket")
        try:
            bucket = self.s3_resource.Bucket(bucket_name)
            print("Successfully listed objects in S3 bucket: ", bucket_name)
            return bucket.objects.filter(Prefix=prefix)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Failed to list objects in S3 bucket: ", err)

File: sample_apps_reinvent2021/go/utils/timestream-dependency-helper.go
package utils

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/service/iam"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/s3/s3manager"
	"github.com/aws/aws-sdk-go/service/sns"
	"github.com/aws/aws-sdk-go/service/sqs"
	"github.com/aws/aws-sdk-go/service/sts"
	"github.com/aws/aws-sdk-go/service/timestreamquery"
	"math/rand"
	"os"
	"runtime"
	"time"
)

/**

 */

const (
	POLICY_DOCUMENT = "{" +
		"  \"Version\": \"2012-10-17\"," +
		"  \"Statement\": [" +
		"    {" +
		"        \"Action\": [" +
		"            \"kms:Decrypt\"," +
		"            \"sns:Publish\"," +
		"            \"timestream:describeEndpoints\"," +
		"            \"timestream:Select\"," +
		"            \"timestream:SelectValues\"," +
		"            \"timestream:WriteRecords\"," +
		"            \"s3:PutObject\"" +
		"       ]," +
		"       \"Resource\": \"*\"," +
		"       \"Effect\": \"Allow\"" +
		"    }" +
		"   ]" +
		"}"

	SQS_POLICY_FORMAT = "{" +
		"\"Version\":\"2012-10-17\"," +
		"\"Statement\": [" +
		"   {" +
		"       \"Sid\":\"topic-subscription-%s\", " +
		"       \"Effect\": \"Allow\", " +
		"       \"Principal\":{\"AWS\":\"*\"}, " +
		"       \"Action\": \"sqs:SendMessage\", " +
		"       \"Resource\": \"%s\", " +
		"       \"Condition\":{" +
		"           \"ArnEquals\":{" +
		"               \"aws:SourceArn\":\"%s\"" +
		"           }" +
		"       }" +
		"   }" +
		"  ]" +
		"}"

	ROLE_POLICY_FORMAT = "{" +
		"\"Version\":\"2012-10-17\"," +
		"\"Statement\":[" +
		"   {" +
		"       \"Effect\":\"Allow\"," +
		"       \"Principal\":{" +
		"            \"Service\": \"timestream.amazonaws.com\"" +
		"       }," +
		"       \"Action\":\"sts:AssumeRole\"" +
		"   }" +
		"  ]" +
		"}"
)

type TimestreamDependencyHelper struct {
	SnsSvc *sns.SNS
	SqsSvc *sqs.SQS
	S3Svc  *s3.S3
	IamSvc *iam.IAM
	StsSvc *sts.STS
}

type NotificationMessage struct {
	ParseFlag         bool
	TopicArn          string `json:"TopicArn"`
	ReceiptHandle     string
	MessageAttributes struct {
		QueryArn struct {
			Type  string `json:"type"`
			Value string `json:"value"`
		} `json:"queryArn"`
		NotificationType struct {
			Type  string `json:"type"`
			Value string `json:"value"`
		} `json:"notificationType"`
	} `json:"MessageAttributes"`
	Type          string `json:"Type"`
	MessageString string `json:"Message"`
	Message       struct {
		Type                      string `json:"type"`
		Arn                       string `json:"arn"`
		NextInvocationEpochSecond int64  `json:"NextInvocationEpochSecond"`
		ScheduledQueryRunSummary  struct {
			FailureReason         string `json:"failureReason"`
			RunStatus             string `json:"runStatus"`
			InvocationEpochSecond int64  `json:"invocationEpochSecond"`
			TriggerTimeMillis     int64  `json:"triggerTimeMillis"`
			ErrorReportLocation   struct {
				S3ReportLocation struct {
					BucketName string `json:"bucketName"`
					ObjectKey  string `json:"objectKey"`
				} `json:"s3ReportLocation"`
				ExecutionStats map[string]interface{} `json:"executionStats"`
			} `json:"errorReportLocation"`
		} `json:"scheduledQueryRunSummary"`
	}
}

type Resource struct {
	Type string
	Identifier  string
	AdditionalDetails string
}
func HandleError(err error, errorMessage string, exitFlag bool) {
	if err != nil {
		_, fn, line, _ := runtime.Caller(1)
		fmt.Printf("[error] %s:%d %s with error %v\n", fn, line, errorMessage, err)
		if exitFlag == true {
			panic(err)
		}
	}

}

func FileExists(name string) (bool, error) {
	_, err := os.Stat(name)
	if err == nil {
		return true, nil
	}
	return false, err
}

func GenerateRandomStringWithSize(size int) string {
	rand.Seed(time.Now().UnixNano())
	alphaNumericList := []rune("abcdefghijklmnopqrstuvwxyz0123456789")
	randomPrefix := make([]rune, size)
	for i := range randomPrefix {
		randomPrefix[i] = alphaNumericList[rand.Intn(len(alphaNumericList))]
	}
	return string(randomPrefix)
}

func parseNotificationMessage(message string) (NotificationMessage, error) {
	var notificationMessage NotificationMessage
	err := json.Unmarshal([]byte(message), &notificationMessage)
	if err != nil {
		HandleError(err, "", false)
		return notificationMessage, err
	}
	err = json.Unmarshal([]byte(notificationMessage.MessageString), &notificationMessage.Message)
	if err != nil {
		HandleError(err, "", false)
		return notificationMessage, err
	}
	notificationMessage.MessageString = ""
	return notificationMessage, err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) CreateSnsTopic(topicName string) (string, error) {
	createTopicInput := &sns.CreateTopicInput{
		Name:       aws.String(topicName),
		Attributes: map[string]*string{"FifoTopic": aws.String("true"), "ContentBasedDeduplication": aws.String("false")},
	}

	createTopicOutput, err := timestreamDependencyHelper.SnsSvc.CreateTopic(createTopicInput)

	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("Creating SNS topic is successful")
		return *createTopicOutput.TopicArn, nil
	}
	return "", err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) CreateSqsQueue(queueName string) (string, error) {

	createQueueInput := &sqs.CreateQueueInput{
		QueueName:  aws.String(queueName),
		Attributes: map[string]*string{"FifoQueue": aws.String("true"), "ContentBasedDeduplication": aws.String("false")},
	}

	createQueueOutput, err := timestreamDependencyHelper.SqsSvc.CreateQueue(createQueueInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Printf("Create SQS Queue with name %s is successful\n", queueName)
		return *createQueueOutput.QueueUrl, nil
	}

	return "", err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) GetSqsQueryUrl(queueName string) (string, error) {

	getQueueUrlInput := &sqs.GetQueueUrlInput{
		QueueName: aws.String(queueName),
	}

	getQueueUrlOutput, err := timestreamDependencyHelper.SqsSvc.GetQueueUrl(getQueueUrlInput)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case timestreamquery.ErrCodeConflictException:
				fmt.Println(timestreamquery.ErrCodeConflictException, aerr.Error())
				return "", nil
			default:
				fmt.Println(aerr.Error())
			}
		} else {
			fmt.Println(err.Error())
		}
		return "", err
	} else {
		return *getQueueUrlOutput.QueueUrl, nil
	}
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteSqsQueue(queueUrl string) error {

	deleteQueueInput := &sqs.DeleteQueueInput{
		QueueUrl: aws.String(queueUrl),
	}

	_, err := timestreamDependencyHelper.SqsSvc.DeleteQueue(deleteQueueInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("delete_sqs_queue is successful")
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) ReceiveMessage(queueUrl string) (NotificationMessage, error) {
	receiveMessageInput := &sqs.ReceiveMessageInput{
		QueueUrl:            aws.String(queueUrl),
		WaitTimeSeconds:     aws.Int64(20),
		MaxNumberOfMessages: aws.Int64(1),
	}
	receiveMessageOutput, err := timestreamDependencyHelper.SqsSvc.ReceiveMessage(receiveMessageInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("received Message successfully")
		messages := receiveMessageOutput.Messages
		if messages != nil && len(messages) > 0 {
			messageBody := messages[0].Body
			response, err := parseNotificationMessage(*messageBody)
			if err == nil {
				response.ReceiptHandle = *messages[0].ReceiptHandle
				response.ParseFlag = true
				return response, nil
			}

		}
	}
	return NotificationMessage{}, err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteMessage(queueUrl string, receiptHandle string) error {
	deleteMessageInput := &sqs.DeleteMessageInput{
		QueueUrl:      aws.String(queueUrl),
		ReceiptHandle: aws.String(receiptHandle),
	}
	_, err := timestreamDependencyHelper.SqsSvc.DeleteMessage(deleteMessageInput)
	if err != nil {
		fmt.Printf("sqs DeleteMessage is failed with error: %s", err.Error())
	} else {
		fmt.Println("sqs DeleteMessage is successful")
	}
	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteSnsTopic(topicArn string) error {

	deleteTopicInput := &sns.DeleteTopicInput{
		TopicArn: aws.String(topicArn),
	}

	_, err := timestreamDependencyHelper.SnsSvc.DeleteTopic(deleteTopicInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("DeleteTopic is successful")
	}
	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) GetSqsQueueArn(queueUrl string) (string, error) {

	createGetQueueAttributesInput := &sqs.GetQueueAttributesInput{
		QueueUrl:       aws.String(queueUrl),
		AttributeNames: []*string{aws.String("QueueArn")},
	}

	createGetQueueAttributesOutput, err := timestreamDependencyHelper.SqsSvc.GetQueueAttributes(createGetQueueAttributesInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Printf("GetQueueAttributes is successful : %s", createGetQueueAttributesOutput.String())
		return *createGetQueueAttributesOutput.Attributes["QueueArn"], nil
	}

	return "", err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) SetSqsAccessPolicy(queueUrl string, topicArn string, queueArn string) error {

	setQueueAttributesInput := &sqs.SetQueueAttributesInput{
		QueueUrl:   aws.String(queueUrl),
		Attributes: map[string]*string{"Policy": aws.String(fmt.Sprintf(SQS_POLICY_FORMAT, topicArn, queueArn, topicArn))},
	}

	setQueueAttributesOutput, err := timestreamDependencyHelper.SqsSvc.SetQueueAttributes(setQueueAttributesInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Printf("setQueueAttributesOutput is successful : %s", setQueueAttributesOutput.GoString())
	}
	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) GetIamRole(roleName string) (string, error) {

	getRoleInput := &iam.GetRoleInput{
		RoleName: aws.String(roleName),
	}

	getRoleOutput, err := timestreamDependencyHelper.IamSvc.GetRole(getRoleInput)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case iam.ErrCodeNoSuchEntityException:
				fmt.Println(iam.ErrCodeNoSuchEntityException, aerr.Error())
				return "", nil
			default:
				fmt.Printf("Error: %s\n", aerr.Error())
			}
		} else {
			fmt.Printf("Error: %s\n", err.Error())
		}
		return "", err
	}

	fmt.Println("getIamRole is successful : %s\n", getRoleOutput.GoString())
	return *getRoleOutput.Role.Arn, nil
}

func (timestreamDependencyHelper TimestreamDependencyHelper) CreateIamRole(roleName string) (string, error) {

	roleArn, err := timestreamDependencyHelper.GetIamRole(roleName)
	if err == nil && roleArn != "" {
		return roleArn, err
	}
	createRoleInput := &iam.CreateRoleInput{
		RoleName:                 aws.String(roleName),
		Description:              aws.String("Created using the AWS SDK for Go"),
		AssumeRolePolicyDocument: aws.String(ROLE_POLICY_FORMAT),
	}

	createRoleOutput, err := timestreamDependencyHelper.IamSvc.CreateRole(createRoleInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Printf("createRoleOutput is successful : %s\n", createRoleOutput.GoString())
		getRoleInput := &iam.GetRoleInput{RoleName: createRoleOutput.Role.RoleName}
		err = timestreamDependencyHelper.IamSvc.WaitUntilRoleExists(getRoleInput)
		if err != nil {
			return "", err
		}
		return *createRoleOutput.Role.Arn, nil
	}

	return "", err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) AttachIamPolicy(roleName string, policyArn string) error {

	attachRolePolicyInput := &iam.AttachRolePolicyInput{
		RoleName:  aws.String(roleName),
		PolicyArn: aws.String(policyArn),
	}

	_, err := timestreamDependencyHelper.IamSvc.AttachRolePolicy(attachRolePolicyInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Printf("attaching RolePolicy is successful\n")
	}

	return err

}

func (timestreamDependencyHelper TimestreamDependencyHelper) DetachIamPolicy(roleName string, policyArn string) error {

	detachRolePolicyInput := &iam.DetachRolePolicyInput{
		RoleName:  aws.String(roleName),
		PolicyArn: aws.String(policyArn),
	}

	_, err := timestreamDependencyHelper.IamSvc.DetachRolePolicy(detachRolePolicyInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("DetachRolePolicy is successful")
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteIamPolicy(policyArn string) error {

	deletePolicyInput := &iam.DeletePolicyInput{
		PolicyArn: aws.String(policyArn),
	}

	_, err := timestreamDependencyHelper.IamSvc.DeletePolicy(deletePolicyInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("DeletePolicy is successful")
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteIamRole(roleName string) error {

	deleteRoleInput := &iam.DeleteRoleInput{
		RoleName: aws.String(roleName),
	}

	_, err := timestreamDependencyHelper.IamSvc.DeleteRole(deleteRoleInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("delete_IAM_role is successful")
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) CreateIamPolicy(policyName string) (string, error) {

	createPolicyInput := &iam.CreatePolicyInput{
		PolicyName:     aws.String(policyName),
		PolicyDocument: aws.String(POLICY_DOCUMENT),
	}

	createPolicyOutput, err := timestreamDependencyHelper.IamSvc.CreatePolicy(createPolicyInput)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case iam.ErrCodeEntityAlreadyExistsException:
				fmt.Println(iam.ErrCodeEntityAlreadyExistsException, aerr.Error())
				return timestreamDependencyHelper.getPolicyArn(policyName), nil
			default:
				fmt.Printf("Error: %s\n", aerr.Error())
			}
		} else {
			fmt.Printf("Error: %s\n", err.Error())
		}
		return "", err
	} else {
		fmt.Printf("CreatePolicy is successful : %s\n", createPolicyOutput.GoString())
		return *createPolicyOutput.Policy.Arn, nil
	}
}

func (timestreamDependencyHelper TimestreamDependencyHelper) SubscribeToSnsTopic(topicArn string, queueArn string) (string, error) {

	createSubscribeInput := &sns.SubscribeInput{
		TopicArn: aws.String(topicArn),
		Protocol: aws.String("sqs"),
		Endpoint: aws.String(queueArn),
	}

	createQueueOutput, err := timestreamDependencyHelper.SnsSvc.Subscribe(createSubscribeInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("Create SQS Queue is successful")
		return *createQueueOutput.SubscriptionArn, nil
	}

	return "", err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) UnsubscribeToSnsTopic(subscriptionArn string) error {

	unsubscribeInput := &sns.UnsubscribeInput{
		SubscriptionArn: aws.String(subscriptionArn),
	}

	_, err := timestreamDependencyHelper.SnsSvc.Unsubscribe(unsubscribeInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("unsubscribe_to_sns_topic is successful")
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) CreateS3Bucket(bucketName string, region string) error {

	fmt.Printf("Creating S3 Bucket with name %s in region %s\n", bucketName, region)

	createBucketInput := &s3.CreateBucketInput{
		Bucket: aws.String(bucketName),
	}
	if region != "us-east-1" {
		createBucketInput.CreateBucketConfiguration = &s3.CreateBucketConfiguration{
			LocationConstraint: aws.String(region),
		}
	}
	createBucketOutput, err := timestreamDependencyHelper.S3Svc.CreateBucket(createBucketInput)

	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("Create S3 Bucket is successful : %v\n", *createBucketOutput)
	}

	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) DeleteS3Bucket(bucketName string) error {

	// Setup BatchDeleteIterator to iterate through a list of objects.
	iter := s3manager.NewDeleteListIterator(timestreamDependencyHelper.S3Svc, &s3.ListObjectsInput{
		Bucket: aws.String(bucketName),
	})

	// Traverse iterator deleting each object
	if err := s3manager.NewBatchDeleteWithClient(timestreamDependencyHelper.S3Svc).Delete(aws.BackgroundContext(), iter); err != nil {
		fmt.Printf("Unable to delete objects from bucket %q, %v\n", bucketName, err)
		return err
	}
	fmt.Printf("Deleted object(s) from bucket: %s\n", bucketName)
	deleteBucketInput := &s3.DeleteBucketInput{
		Bucket: aws.String(bucketName),
	}
	_, err := timestreamDependencyHelper.S3Svc.DeleteBucket(deleteBucketInput)

	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
	} else {
		fmt.Println("delete S3 Bucket is successful")
	}
	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) getAccountId() (string, error) {

	getCallerIdentityInput := &sts.GetCallerIdentityInput{}
	getCallerIdentityOutput, err := timestreamDependencyHelper.StsSvc.GetCallerIdentity(getCallerIdentityInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
		return "", nil
	} else {
		fmt.Printf("get_account_id is successful : %s\n", getCallerIdentityOutput)
		return *getCallerIdentityOutput.Account, nil
	}
}

func (timestreamDependencyHelper TimestreamDependencyHelper) getPolicyArn(policyName string) string {
	accountId, _ := timestreamDependencyHelper.getAccountId()
	return fmt.Sprintf("arn:aws:iam::%s:policy/%s", accountId, policyName)
}

func (timestreamDependencyHelper TimestreamDependencyHelper) listReportFilesInS3(s3ErrorReportBucketName string,
	errorReportPrefix string) ([]*s3.Object, error) {
	listObjectsInput := &s3.ListObjectsInput{
		Bucket: aws.String(s3ErrorReportBucketName),
		Prefix: aws.String(errorReportPrefix),
	}
	listObjectsOutput, err := timestreamDependencyHelper.S3Svc.ListObjects(listObjectsInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
		return nil, nil
	} else {
		fmt.Printf("listObjectsOutput is successful : %s\n", listObjectsOutput)
		return listObjectsOutput.Contents, nil
	}
}

func (timestreamDependencyHelper TimestreamDependencyHelper) ParseS3ErrorReport(s3ErrorReportBucketName string,
	errorReportPrefix string) error {
	s3Objects, err := timestreamDependencyHelper.listReportFilesInS3(s3ErrorReportBucketName, errorReportPrefix)
	if err != nil {
		return err
	}
	fmt.Printf("s3Objects first object contents : %+v\n", s3Objects[0])
	err = timestreamDependencyHelper.readErrorReport(s3Objects[0], s3ErrorReportBucketName)
	return err
}

func (timestreamDependencyHelper TimestreamDependencyHelper) readErrorReport(s3Object *s3.Object,
	s3ErrorReportBucketName string) error {
	getObjectInput := &s3.GetObjectInput{
		Key:    s3Object.Key,
		Bucket: aws.String(s3ErrorReportBucketName),
	}
	getObjectOutput, err := timestreamDependencyHelper.S3Svc.GetObject(getObjectInput)
	if err != nil {
		fmt.Printf("Error: %s\n", err.Error())
		return err
	}
	buf := new(bytes.Buffer)
	buf.ReadFrom(getObjectOutput.Body)
	objectContents := buf.String()
	fmt.Printf("S3 Error Report Contents : %+v\n", objectContents)
	return nil
}
